lib/logstash/outputs/elasticsearch/ilm.rb

```diff
@@ -14,11 +14,18 @@ def ilm_in_use?
       return @ilm_actually_enabled if defined?(@ilm_actually_enabled)
       @ilm_actually_enabled =
         begin
-          if @ilm_enabled == 'auto'
+          if serverless?
+            raise LogStash::ConfigurationError, "Invalid ILM configuration `ilm_enabled => true`. " +
```

Comment:
Unfortunately with this error the pipeline doesn't terminate, and loops with the error:

```
[2023-08-07T13:48:10,631][INFO ][logstash.outputs.elasticsearch][main] Using a default mapping template {:es_version=>8, :ecs_compatibility=>:v8}
[2023-08-07T13:48:10,633][ERROR][logstash.outputs.elasticsearch][main] Failed to install template {:message=>"Invalid ILM configuration `ilm_enabled => true`. Serverless Elasticsearch cluster does not support Index Lifecycle Management.", :exception=>LogStash::ConfigurationError, :backtrace=>["/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/outputs/elasticsearch/ilm.rb:18:in `ilm_in_use?'", "/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/outputs/elasticsearch/template_manager.rb:30:in `install_template'", "/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/outputs/elasticsearch.rb:595:in `install_template'", "/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/outputs/elasticsearch.rb:353:in `finish_register'", "/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/outputs/elasticsearch.rb:326:in `block in register'", "/Users/joaoduarte/elastic/logstash-plugins/logstash-output-elasticsearch/lib/logstash/plugin_mixins/elasticsearch/common.rb:172:in `block in after_successful_connection'"]}
[2023-08-07T13:48:10,844][INFO ][logstash.javapipeline    ][main] Pipeline Java execution initialization time {"seconds"=>0.26}
[2023-08-07T13:48:10,855][INFO ][logstash.javapipeline    ][main] Pipeline started {"pipeline.id"=>"main"}
The stdin plugin is now waiting for input:
[2023-08-07T13:48:10,866][INFO ][logstash.agent           ] Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
[2023-08-07T13:48:10,900][ERROR][logstash.outputs.elasticsearch][main][d55e1ed556e8769d7d33259016cd16b02571398e3b6a56bce084da16b6ea00ea] Elasticsearch setup did not complete normally, please review previously logged errors {:message=>"Invalid ILM configuration `ilm_enabled => true`. Serverless Elasticsearch cluster does not support Index Lifecycle Management.", :exception=>LogStash::ConfigurationError}
[2023-08-07T13:48:15,886][ERROR][logstash.outputs.elasticsearch][main][d55e1ed556e8769d7d33259016cd16b02571398e3b6a56bce084da16b6ea00ea] Elasticsearch setup did not complete normally, please review previously logged errors {:message=>"Invalid ILM configuration `ilm_enabled => true`. Serverless Elasticsearch cluster does not support Index Lifecycle Management.", :exception=>LogStash::ConfigurationError}
[2023-08-07T13:48:20,862][ERROR][logstash.outputs.elasticsearch][main][d55e1ed556e8769d7d33259016cd16b02571398e3b6a56bce084da16b6ea00ea] Elasticsearch setup did not complete normally, please review previously logged errors {:message=>"Invalid ILM configuration `ilm_enabled => true`. Serverless Elasticsearch cluster does not support Index Lifecycle Management.", :exception=>LogStash::ConfigurationError}
[2023-08-07T13:48:25,872][ERROR][logstash.outputs.elasticsearch][main][d55e1ed556e8769d7d33259016cd16b02571398e3b6a56bce084da16b6ea00ea] Elasticsearch setup did not complete normally, please review previously logged errors {:message=>"Invalid ILM configuration `ilm_enabled => true`. Serverless Elasticsearch cluster does not support Index Lifecycle Management.", :exception=>LogStash::ConfigurationError}
```

Reply:
We have two behaviors on errors in template and ilm. 
1.  The plugin [logs](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v11.15.9/lib/logstash/outputs/elasticsearch.rb#L597) error message when it fails to install template, such as failing to [read](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v11.15.9/lib/logstash/outputs/elasticsearch/template_manager.rb#L97) the template file. 
2.  The plugin loops with error when it fails ilm [check](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v11.15.9/lib/logstash/outputs/elasticsearch/ilm.rb#L58), such as policy not found.

This change keeps the same way to handle error, so now my question is do we want to introduce a new error behavior to terminate the current pipeline? Do we want to unify the errors in template and ilm to stop the pipeline when template file or ilm policy is missing?

Reply:
filed an issue for desired error behaviour https://github.com/logstash-plugins/logstash-output-elasticsearch/issues/1146